### PR TITLE
Add typing to certbot.apache

### PR DESCRIPTION
**Pull request by atombrella**
I noticed that Python 2.7 support has been dropped from that module.

There are still things I don't really know how to do with `typing`, e.g., reference something from the module itself. `obj.Addr` is one example.

There are some irrelevant changes, that probably should go in separate PRs. f-string and the single `object` thing I removed.

## Pull Request Checklist

- [x] If the change being made is to a [distributed component](https://certbot.eff.org/docs/contributing.html#code-components-and-layout), edit the `master` section of `certbot/CHANGELOG.md` to include a description of the change being made.
- [x] Add or update any documentation as needed to support the changes in this PR.
- [x] Include your name in `AUTHORS.md` if you like.

### CODE REVIEW

**Comment by adferrand** [comment]
@atombrella stay tuned on this PR. As soon as my PRs to fully type the `certbot` package, I will go back to your PR so that we implement the full typing for `certbot-apache`.

**Comment by atombrella** [comment]
@adferrand Good :) I just pushed some updates. I'm currently not able to run `tox` locally (it fails with the `augeas` package).

File: certbot-apache/certbot_apache/_internal/obj.py [hunk 1]
```diff
@@ -147,24 +155,17 @@ def get_names(self):
...
-    def display_repr(self):
+            f"File: {self.filep}\n"
+            f"Vhost path: {self.path}\n"
+            f"Addresses: {', '.join(str(addr) for addr in self.addrs)}\n"
+            f"Name: {self.name if self.name is not None else ''}\n"
+            f"Aliases: {', '.join(name for name in self.aliases)}\n"
+            f"TLS Enabled: {'Yes' if self.ssl else 'No'}\n"
+            f"Site Enabled: {'Yes' if self.enabled else 'No'}\n"
+            f"mod_macro Vhost: {'Yes' if self.modmacro else 'No'}"
+        )
```
Comment (atombrella) [inline certbot-apache/certbot_apache/_internal/obj.py:166]: I know this is unrelated to the typing. I can move it to a new PR?

**Comment by atombrella** [comment]
@adferrand I've added a bunch more typing locally, and will try to get the tests to pass. Maybe we can merge this before Christmas :tada: It's rather tempting to do more clean-up in `certbot-apache` while doing this, e.g., use f-strings.

**Comment by adferrand** [comment]
Hello @atombrella, I checked on my side the amount of potential work to type `certbot-apache`, and I also take into consideration the work required to type `certbot-nginx` in https://github.com/certbot/certbot/pull/9124: there is a tremendous amount of work.

So I would prefer that we separate the typing itself and the improvement of the code quality itself. So I can make several propositions for a strategy:
* I can do the typing since I did it for the rest of the code and you focus on the code quality in a separate PR
* you can continue on this PR and focus on typing alone, and then another PR for the code quality

What do you think?

**Comment by adferrand** [comment]
I just looked at the latest commits, you added a lot of types. It would be a shame to loose this PR. So my favor is on continue to focus on the typing here and take the code quality later.

**Comment by atombrella** [comment]
> Hello @atombrella, I checked on my side the amount of potential work to type `certbot-apache`, and I also take into consideration the work required to type `certbot-nginx` in #9124: there is a tremendous amount of work.
> 
> So I would prefer that we separate the typing itself and the improvement of the code quality itself. So I can make several propositions for a strategy:
> 
>     * I can do the typing since I did it for the rest of the code and you focus on the code quality in a separate PR

I have a bunch more stuff locally that I'll push. You have access to my branch, so it's possible I'll request some active help at some point. Annotating with types does indeed take a lot of time, especially fixing the errors/warnings from `mypy`.
 
>     * you can continue on this PR and focus on typing alone, and then another PR for the code quality

Yes. That's also why I opened #9114 which is now merged. Initially this PR was spawned from some contemplation to do TLS 1.3 only support, as [Mozilla SSL configurator](https://ssl-config.mozilla.org/#server=apache&version=2.4.41&config=modern&openssl=1.1.1k&guideline=5.6) has for the `Modern` option, but decided typing was also fun to do.

**Comment by atombrella** [comment]
@adferrand I added loads more types today. I think it's ready to be reviewed, and looked over :) I use PyCharm, which still complains a bit a few places, but locally, `mypy` and `tox` seem happy. Please note that this is my first major `typing` project, so there are undoubtedly things to improve.

File: certbot-apache/certbot_apache/_internal/apache_util.py [hunk 2]
```diff
@@ -111,12 +117,12 @@ def parse_define_file(filepath, varname):
...
 
 
-def unique_id():
+def unique_id() -> str:
     """ Returns an unique id to be used as a VirtualHost identifier"""
     return binascii.hexlify(os.urandom(16)).decode("utf-8")
 
 
-def included_in_paths(filepath, paths):
+def included_in_paths(filepath: str, paths: List[pathlib.Path]) -> bool:
```
Comment (adferrand) [inline certbot-apache/certbot_apache/_internal/apache_util.py:125]: I think `path` should be `List[str]` here.

File: certbot-apache/certbot_apache/_internal/configurator.py [hunk 3]
```diff
@@ -2497,19 +2497,22 @@ def more_info(self):
                 version=".".join(str(i) for i in self.version))
         )
 
-    def auth_hint(self, failed_achalls): # pragma: no cover
+    # pragma: no cover
```
Comment (adferrand) [inline certbot-apache/certbot_apache/_internal/configurator.py:2500]: I think this change makes the cover ignore directive a noop.

File: certbot-apache/certbot_apache/_internal/configurator.py [hunk 4]
```diff
@@ -2497,19 +2497,22 @@ def more_info(self):
...
     ###########################################################################
-    def get_chall_pref(self, unused_domain):
+    def get_chall_pref(self, unused_domain: str) -> List[Type[challenges.HTTP01]]:
         """Return list of challenge preferences."""
         return [challenges.HTTP01]
 
-    def perform(self, achalls):
+    def perform(
+        self, achalls: List[KeyAuthorizationAnnotatedChallenge]
+    ) -> List[Optional[challenges.HTTP01Response]]:
```
Comment (adferrand) [inline certbot-apache/certbot_apache/_internal/configurator.py:2515]: Given the interface, `perform` should return `List[challenges.HTTP01Response]`

File: certbot-apache/certbot_apache/_internal/configurator.py [hunk 5]
```diff
@@ -2542,14 +2545,14 @@ def perform(self, achalls):
 
         return responses
 
-    def _update_responses(self, responses, chall_response, chall_doer):
+    def _update_responses(self, responses: List, chall_response, chall_doer) -> None:
```
Comment (adferrand) [inline certbot-apache/certbot_apache/_internal/configurator.py:2548]: You can precise `List[challenges.HTTP01Response]` for `responses`.

**Review by adferrand — CHANGES_REQUESTED** [review]
Thanks a lot @atombrella, here is my review. I will do a second round once all checks pass.

File: certbot-apache/certbot_apache/_internal/augeasparser.py [hunk 6]
```diff
@@ -203,7 +210,7 @@ def __eq__(self, other):
                     self.metadata == other.metadata)
         return False
 
-    def set_parameters(self, parameters):
+    def set_parameters(self, parameters: List[str]):
```
[the remark this answers is not in the document]
Comment (atombrella) [reply certbot-apache/certbot_apache/_internal/augeasparser.py:213]: https://docs.python.org/3/library/typing.html#typing.List Ah, I missed this :) `List` and `Iterable` are easy to mix up.

File: certbot-apache/certbot_apache/_internal/apache_util.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (atombrella) [reply certbot-apache/certbot_apache/_internal/apache_util.py:125]: Just to follow https://docs.python.org/3/library/typing.html#typing.List I'll do `Iterable`.

File: certbot-apache/certbot_apache/_internal/parsernode_util.py [hunk 7]
```diff
@@ -90,7 +95,7 @@ def commentnode_kwargs(kwargs):
     return comment, kwargs
 
 
-def directivenode_kwargs(kwargs):
+def directivenode_kwargs(kwargs: Dict[str, Any]) -> Tuple[Any, Any, Any, Dict]:
```
[the remark this answers is not in the document]
Comment (atombrella) [reply certbot-apache/certbot_apache/_internal/parsernode_util.py:98]: I get a bunch of warnings when using `Mapping` :/

**Comment by atombrella** [comment]
@adferrand While going through your remarks, I'm noticing some odd things in the code. One thing in particular is the mismatch with error handling, and returns. One good example is `def _create_vhost(self, path: str) -> Optional[obj.VirtualHost]:` which causes various warnings, because `make_vhost_ssl` in the `Configurator` has `ssl_vhost = self._create_vhost(vh_p)` which then expects a non-null object. There are still failures locally that I'm working my way through.

I'm noting down this as a future improvement idea: keyword-only arguments in different places of the module. It'll make some things easier and safer when passing arguments around. Python 3 introduced them in https://www.python.org/dev/peps/pep-3102/ Personally I enjoy them, and while recently getting a bit into GoLang, I was discussing with a previous colleague how useful this is in Python.

File: certbot-apache/certbot_apache/_internal/http_01.py [hunk 8]
```diff
@@ -64,7 +67,7 @@ def __init__(self, configurator: "ApacheConfigurator") -> None:
             "http_challenges")
         self.moded_vhosts: Set[VirtualHost] = set()
 
-    def perform(self):
+    def perform(self) -> List[Challenge]:
```
[the remark this answers is not in the document]
Comment (atombrella) [reply certbot-apache/certbot_apache/_internal/http_01.py:70]: It's getting the data from `_set_up_challenges`, which returns `List[Challenge]`.

**Comment by atombrella** [comment]
@adferrand I fixed some of the warnings. I'd like some active assistance to pushing to over the finish line. Currently there are some 30-something warnings.

**Comment by atombrella** [comment]
```
certbot-apache/certbot_apache/_internal/parser.py:97: error: Item "None" of "Optional[ApacheConfigurator]" has no attribute "options"  [union-attr]
certbot-apache/certbot_apache/_internal/parser.py:166: error: Item "None" of "Optional[ApacheConfigurator]" has no attribute "save_notes"  [union-attr]
certbot-apache/certbot_apache/_internal/parser.py:191: error: Item "None" of "Optional[ApacheConfigurator]" has no attribute "save_notes"  [union-attr]
certbot-apache/certbot_apache/_internal/parser.py:192: error: Item "None" of "Optional[ApacheConfigurator]" has no attribute "save"  [union-attr]
certbot-apache/certbot_apache/_internal/parser.py:203: error: Item "None" of "Optional[ApacheConfigurator]" has no attribute "save_notes"  [union-attr]
certbot-apache/certbot_apache/_internal/parser.py:225: error: Item "None" of "Optional[ApacheConfigurator]" has no attribute "save_notes"  [union-attr]
certbot-apache/certbot_apache/_internal/parser.py:297: error: Item "None" of "Optional[ApacheConfigurator]" has no attribute "options"  [union-attr]
certbot-apache/certbot_apache/_internal/parser.py:307: error: Item "None" of "Optional[ApacheConfigurator]" has no attribute "options"  [union-attr]
certbot-apache/certbot_apache/_internal/parser.py:316: error: Item "None" of "Optional[ApacheConfigurator]" has no attribute "options"  [union-attr]
```

This is one part of my heacaches with annotating the Apache plugin. There are loads of `return None` where I think it'd be more correct to throw an error instead.

@bmw @alexzorin @adferrand Could you please add the labels to this PR? `area: python`, `area: apache`.

**Comment by adferrand** [comment]
Hello @atombrella. First thing first, I will be available again next week to review the state of the work you did here. But it seems that most of the issues are fixed now, we are almost there!

For the remaining issues, yes I have encountered a similar situation with the Nginx plugin. Basically, all there errors mean that if the plugin is not "prepared", it will not work. What is unfortunate with the current design is that this preparation is not part of the real Python object initialization, but is more like a lazy process called at a later time using the `prepare()` method. From what I understood it is done like this to be able to manipulate a reference of the plugin before it is actually used (eg for plugin selection).

This is something that `mypy` cannot understand, since there are higher level process ensuring that the actual logic of the plugin is not called before the plugin is prepared. We will have to help `mypy` by writing some exceptions. In the Nginx plugin, I tried to limit these exception as much as possible by writing them early in the plugin lifecycle so that we can suppose also very early that the critical variables are not `None` anymore.

**Comment by adferrand** [comment]
@atombrella, some conflicts here since your cleanup in the Apache plugin code.

**Comment by atombrella** [comment]
@adferrand rebased and fixed the conflicts in case you haven't noticed.

**Comment by adferrand** [comment]
@atombrella as requested, I pushed some commits on your PR to solve the remaining errors.

**Comment by atombrella** [comment]
@adferrand Thank you :) As you wrote, it doesn't seem like a massive amount of stuff you had to change :tada: 

When you merge, could you please edit the commit message inside the merge dialog so it's not full of the intermediate commit messages (obviously, just remember the `Co-Authored-By: ..`) ? It clutters `git blame`, and it's referenced in the PR.

**Review by adferrand — APPROVED** [review]
Full test campaign is passing. Types looks great. LGTM!

**Comment by adferrand** [comment]
@atombrella I reduced the list of commits to only one significant, as you requested.

**Pull request merged**